File: src/engine/contingency.ts
import { runPowerFlow } from './powerflow';
import { BRANCHES, BRANCH_MAP, BUS_MAP, LOAD_BUS_IDS } from './systemModel';
import type { OperatingState } from './types';

// ─────────────────────────────────────────────────────────────────────────────
//  N-1 상정고장 스크리닝
//  현재 투입 중인 각 선로를 하나씩 탈락시켜 계통 안전성을 평가한다.
// ─────────────────────────────────────────────────────────────────────────────

export interface ContingencyCase {
  branchId: number;
  branchName: string;
  converged: boolean;
  islanded: boolean; // 탈락으로 고립(정전) 모선 발생 여부
  islandedBuses: number[];
  lostLoadMW: number; // 탈락으로 정전된 부하량
  lostGenOnly: boolean; // 발전만 탈락(부하 정전 없음)
  maxLoadingPct: number; // 잔여 계통 최대 선로 부하율
  worstBranchId: number | null;
  anyOverload: boolean;
  severity: 'safe' | 'warning' | 'critical';
}

export interface ContingencyReport {
  cases: ContingencyCase[];
  secure: boolean; // 모든 N-1 케이스가 안전(safe)한가
}

export function runContingencyAnalysis(
  state: OperatingState,
): ContingencyReport {
  const cases: ContingencyCase[] = [];

  for (const br of BRANCHES) {
    if (!state.breakers[br.id]) continue; // 이미 개방된 선로는 대상 외

    const trial: OperatingState = {
      ...state,
      breakers: { ...state.breakers, [br.id]: false },
    };
    const res = runPowerFlow(trial);

    let maxLoadingPct = 0;
    let worstBranchId: number | null = null;
    for (const r of Object.values(res.branches)) {
      if (r.energized && r.loadingPct > maxLoadingPct) {
        maxLoadingPct = r.loadingPct;
        worstBranchId = r.id;
      }
    }

    const islanded = res.islandedBuses.length > 0;
    const lostLoadBuses = res.islandedBuses.filter((id) =>
      LOAD_BUS_IDS.includes(id),
    );
    const lostLoadMW = lostLoadBuses.reduce(
      (s, id) => s + (state.loads[id] ?? 0),
      0,
    );
    const lostGenOnly =
      islanded &&
      lostLoadBuses.length === 0 &&
      res.islandedBuses.every((id) => BUS_MAP[id].category === 'generator');

    // 심각도 분류:
    //  critical → 미수렴 또는 부하 정전(load shedding) 발생
    //  warning  → 발전만 탈락(slack 보상) 또는 잔여 선로 과부하
    //  safe     → 무영향
    let severity: ContingencyCase['severity'] = 'safe';
    if (!res.converged || lostLoadMW > 0) {
      severity = 'critical';
    } else if (lostGenOnly || res.anyOverload) {
      severity = 'warning';
    }

    cases.push({
      branchId: br.id,
      branchName: BRANCH_MAP[br.id].name,
      converged: res.converged,
      islanded,
      islandedBuses: res.islandedBuses,
      lostLoadMW,
      lostGenOnly,
      maxLoadingPct,
      worstBranchId,
      anyOverload: res.anyOverload,
      severity,
    });
  }

  return {
    cases,
    secure: cases.every((c) => c.severity === 'safe'),
  };
}

File: python/contingency.py
"""
N-1 해석 — 선로·변압기를 하나씩 개방하며 PF 반복 실행
"""

from powerflow import run_power_flow
from system_model import BRANCHES, DEFAULT_STATE
import copy


def run_n1_analysis(state: dict = None) -> list:
    """
    모든 선로에 대해 N-1 해석 수행
    반환: 각 선로 개방 시 결과 리스트
    """
    if state is None:
        state = DEFAULT_STATE

    results = []
    for br in BRANCHES:
        # 해당 선로 차단기만 개방
        test_state = copy.deepcopy(state)
        test_state['breakers'][br['id']] = False

        pf = run_power_flow(test_state)
        results.append({
            'branchId':   br['id'],
            'branchName': br['name'],
            'converged':  pf['converged'],
            'islanded':   pf['islandedBuses'],
            'totalGenMW': pf['totalGenMW'],
            'totalLossMW':pf['totalLossMW'],
            'buses':      pf['buses'],
            'branches':   pf['branches'],
            'overloads':  [b['name'] for b in BRANCHES
                           if pf['branches'].get(b['id'], {}).get('overloaded')],
        })
    return results
